framework/src/org/checkerframework/framework/flow/CFAbstractValue.java

```diff
@@ -4,550 +4,481 @@
+ * {@link AnnotatedTypeVariable} is computed using the type mirror.  For wildcards, it is not
+ * always possible to get the {@link AnnotatedWildcardType} for the type mirror.  However, a
+ * CFAbstractValue's type mirror is only a wildcard if the type of some expression is a wildcard,
+ * so it is always the use of a wildcard rather than the declaration. For uses of wildcards, only
+ * the primary annotation on the upper bound is ever used.  So, the set of annotations represents
+ * the primary annotation on the wildcard's upper bound.  If that upper bound is a type variable,
+ * then the set of annotations could be missing an annotation in a hierarchy.
  *
+ * @author Stefan Heule
  */
 public abstract class CFAbstractValue<V extends CFAbstractValue<V>> implements AbstractValue<V> {
 
     /**
-     * The analysis class this store belongs to.
+     * The analysis class this value belongs to.
      */
     protected final CFAbstractAnalysis<V, ?, ?> analysis;
 
     protected final TypeHierarchy typeHierarchy;
+    protected final GenericAnnotatedTypeFactory<V, ?, ?, ?> factory;
+    protected final ProcessingEnvironment processingEnv;
+    protected final QualifierHierarchy hierarchy;
+    protected final TypeMirror underlyingType;
+    protected final Set<AnnotationMirror> annotations;
+
+    public CFAbstractValue(
+            CFAbstractAnalysis<V, ?, ?> analysis,
+            Set<AnnotationMirror> annotations,
+            TypeMirror underlyingType) {
+        this.analysis = analysis;
+        this.typeHierarchy = analysis.getTypeHierarchy();
+        this.annotations = annotations;
+        this.underlyingType = underlyingType;
+        this.factory = analysis.getTypeFactory();
+        this.processingEnv = factory.getProcessingEnv();
+        this.hierarchy = analysis.getTypeFactory().getQualifierHierarchy();
 
-    /**
-     * The type (with annotations) corresponding to this abstract value.
-     */
-    protected final AnnotatedTypeMirror type;
+        validateSet(this.getAnnotations(), this.getUnderlyingType());
+    }
 
     public CFAbstractValue(CFAbstractAnalysis<V, ?, ?> analysis, AnnotatedTypeMirror type) {
-        assert AnnotatedTypes.isValidType(analysis.qualifierHierarchy, type)
-                : "Encountered invalid type: " + type.toString(true);
+        if (type.getKind() == TypeKind.WILDCARD) {
+            // For wildcards, store the primary annotation of the upper bound.  See the class
+            // comment for more details.
+            AnnotatedWildcardType wildcard = (AnnotatedWildcardType) type;
+            this.annotations = wildcard.getExtendsBound().getAnnotations();
+        } else {
+            this.annotations = type.getAnnotations();
+        }
+
         this.analysis = analysis;
-        this.type = type;
         this.typeHierarchy = analysis.getTypeHierarchy();
-    }
+        this.underlyingType = type.getUnderlyingType();
```

Comment:
This seems to be an auxiliary constructor. You could avoid code duplication by calling the primary constructor.